native/c/zdsm.c

```diff
@@ -85,3 +89,152 @@ typedef int (*IGWASMS)(
     ZSMS_DATA sms_data[3],
     int *ds_type // 1 for PDSE, 2 for HFS
     ) ATTRIBUTE(amode31);
+
+#pragma prolog(ZDSRECFM, " ZWEPROLG NEWDSA=(YES,128) ")
+#pragma epilog(ZDSRECFM, " ZWEEPILG ")
+
+#ifndef MAX_DSCBS
+#define MAX_DSCBS 12
+#endif
+
+// Obtain the record format for a data set, given its name and volser
+// https://www.ibm.com/docs/en/SSLTBW_2.2.0/pdf/dgt3s310.pdf
+int ZDSRECFM(ZDS *zds, const char *dsn, const char *volser, char *recfm_buf,
+             int recfm_buf_len)
+{
+  int rc = RTNCD_FAILURE;
+  char dsn_upper[45];
+  char vol_upper[7];
+  memset(dsn_upper, 0x40, sizeof(dsn_upper));
+  memset(vol_upper, 0x40, sizeof(vol_upper));
+
+  struct DSCBFormat1 *dscb = NULL;
+  // workarea: 140-byte array (0x8C) * total possible DSCBs to iterate
+  char workarea[0x8C * MAX_DSCBS] = {0};
+  struct ObtainCamlstSearchParams params = {0};
+  memset(&params, 0, sizeof(params));
+  memset(&dscb, 0, sizeof(dscb));
+
+  int i;
+  // Prepare the data set name and volume serial in uppercase for comparison
+  for (i = 0; i < 44 && dsn[i]; i++)
+  {
+    dsn_upper[i] = toupper(dsn[i]);
+  }
+
+  for (i = 0; i < 6 && volser[i]; i++)
+  {
+    vol_upper[i] = toupper(volser[i]);
+  }
+
+  // We're using OBTAIN through CAMLST SEARCH to get the DSCBs, see here for more info:
+  // https://www.ibm.com/docs/en/zos/3.1.0?topic=obtain-reading-dscb-by-data-set-name
+
+  // OBTAIN by data set name
+  params.function_code = (unsigned char)0xC1u;
+  params.number_dscbs = (unsigned char)MAX_DSCBS;
+  // Allow lookup of format-1 or format-8 DSCB
+  params.option_flags = OPTION_EADSCB;
+  params.dsname_ptr = dsn_upper;
+  params.volume_ptr = vol_upper;
+  params.workarea_ptr = &workarea;
+#if defined(__IBM_METAL__)
```

Comment:
also note this is a WIP, but to capture this since we don't have this approach written:
* the `__asm` portion should use the conditional assembly macro instead of manual linkage, e.g. `OBTAIN` instead of `SVC 27`
* the `__asm` portion should be implemented as a C macro (`#define`) in a header
* that C macro is that wrapped in a `static` function in the header (for type checking)